packages/astro/components/Shiki.js

```diff
@@ -0,0 +1,22 @@
+import { getHighlighter as getShikiHighlighter } from 'shiki';
+
+// Caches Promise<Highligher> for reuse when the same theme and langs are provided
+const _resolvedHighlighters = new Map();
+
+function stringify(opts) {
+    // Always sort keys before stringifying to make sure objects match regardless of parameter ordering
+    return JSON.stringify(opts, Object.keys(opts).sort());
+}
+
+export function getHighlighter(opts) {
+    const key = stringify(opts);
+
+    // Highlighter has already been requested, reuse the same instance
+    if (_resolvedHighlighters.has(key)) { return _resolvedHighlighters.get(key) }
+
+    // Start the async getHighlighter call and cache the Promise
```

Comment:
Did a bit more poking around and confirmed that this looks like the memory issue was just due to pages being built in parallel.  Each page initialized a highlighter and Node couldn't garbage collect them in between page builds

Caching the promise here makes sure that each page gets an immediate response and they can all share the same highlighter instance

Reply:
ah nice, we had to do this in the remark plugin also, wish we could reuse more logic between teh two!